File: OneDrive - umindanao.edu.ph/Documents/PAOLO/Python/Expense_tracker.py

from datetime import datetime
from collections import defaultdict


# ─────────────────────────────────────────────
#  Data Classes
# ─────────────────────────────────────────────

class Expense:
    # initialize with amount, category, description, date (optional), and expense ID

    def __init__(self, amount: float, category: str, description: str,
                 date: str = None, expense_id: int = None):
        self.expense_id = expense_id
        self.amount = round(float(amount), 2)
        self.category = category.strip().title()
        self.description = description.strip()
        self.date = date or datetime.now().strftime("%Y-%m-%d")
    # represent a human readable string of the expense for debugging and display purposes
    def __repr__(self):
        return (f"Expense(id={self.expense_id}, ${self.amount:.2f}, "
                f"'{self.category}', '{self.description}', {self.date})")


class Budget:
    # initialize with category and limit amount
    def __init__(self, category: str, limit: float):
        self.category = category.strip().title()
        self.limit = round(float(limit), 2)
    # represent a human readable string of the budget for debugging and display purposes
    def __repr__(self):
        return f"Budget('{self.category}', limit=${self.limit:.2f})"


#  Core Manager

class ExpenseTracker:
    # initialize with empty expenses list, budgets dict, and next ID counter
    def __init__(self):
        self._expenses: list[Expense] = []
        self._budgets: dict[str, Budget] = {}
        self._next_id: int = 1

    # Expenses CRUD

    def add_expense(self, amount: float, category: str,
                    description: str, date: str = None) -> Expense:
        expense = Expense(amount, category, description, date, self._next_id)
        self._next_id += 1
        self._expenses.append(expense)
        return expense

    def get_expense(self, expense_id: int) -> Expense | None:
        for e in self._expenses:
            if e.expense_id == expense_id:
                return e
        return None
# kwargs define every possible parameter upfront
# kwargs can use any of the following keys: amount, category, description, date
    def edit_expense(self, expense_id: int, **kwargs) -> bool:
        expense = self.get_expense(expense_id)
        if not expense:
            return False
        for key, val in kwargs.items():
            if key == "amount":
                expense.amount = round(float(val), 2)
            elif key == "category":
                expense.category = str(val).strip().title()
            elif key == "description":
                expense.description = str(val).strip()
            elif key == "date":
                expense.date = val
        return True

    def delete_expense(self, expense_id: int) -> bool:
        before = len(self._expenses)
        self._expenses = [e for e in self._expenses if e.expense_id != expense_id]
        return len(self._expenses) < before

    def list_expenses(self, category: str = None, month: str = None,
                      sort_by: str = "date", ascending: bool = False) -> list[Expense]:
        result = self._expenses
        if category:
            result = [e for e in result if e.category == category.strip().title()]
        if month:
            result = [e for e in result if e.date.startswith(month)]
        key = (lambda e: e.amount) if sort_by == "amount" else (lambda e: e.date)
        return sorted(result, key=key, reverse=not ascending)

    # ── Budgets ───────────────────────────────

    def set_budget(self, category: str, limit: float) -> Budget:
        budget = Budget(category, limit)
        self._budgets[budget.category] = budget
        return budget

    def get_budget(self, category: str) -> Budget | None:
        return self._budgets.get(category.strip().title())

    def remove_budget(self, category: str) -> bool:
        key = category.strip().title()
        if key in self._budgets:
            del self._budgets[key]
            return True
        return False

    # ── Reports ───────────────────────────────

    def summary(self, month: str = None) -> dict:
        """Return spending summary, optionally filtered by month."""
        expenses = self.list_expenses(month=month)
        total = sum(e.amount for e in expenses)
        by_category = defaultdict(float)
        for e in expenses:
            by_category[e.category] += e.amount

        budget_status = {}
        for cat, spent in by_category.items():
            budget = self._budgets.get(cat)
            budget_status[cat] = {
                "spent": round(spent, 2),
                "limit": budget.limit if budget else None,
                "remaining": round(budget.limit - spent, 2) if budget else None,
                "over_budget": spent > budget.limit if budget else False,
            }

        return {
            "period": month or "All Time",
            "total": round(total, 2),
            "count": len(expenses),
            "by_category": dict(budget_status),
        }

    @property
    def categories(self) -> list[str]:
        return sorted({e.category for e in self._expenses})



# Interface


class ExpenseTrackerInterface:

    MENU = """
╔══════════════════════════════════╗
║       EXPENSE TRACKER            ║
╠══════════════════════════════════╣
║  1. Add Expense                  ║
║  2. View Expenses                ║
║  3. Edit Expense                 ║
║  4. Delete Expense               ║
║  5. Set Budget                   ║
║  6. View Summary / Budget Status ║
║  7. Remove Budget                ║
║  0. Exit                         ║
╚══════════════════════════════════╝"""

    def __init__(self):
        self.tracker = ExpenseTracker()

    def run(self):
        print("\nWelcome to Expense Tracker!")
        while True:
            print(self.MENU)
            choice = input("Choose an option: ").strip()
            actions = {
                "1": self._add_expense,
                "2": self._view_expenses,
                "3": self._edit_expense,
                "4": self._delete_expense,
                "5": self._set_budget,
                "6": self._view_summary,
                "7": self._remove_budget,
                "0": self._exit,
            }
            action = actions.get(choice)
            if action:
                action()
            else:
                print("❌ Invalid option. Try again.")

    # ── Menu Actions ──────────────────────────

    def _add_expense(self):
        print("\n── Add Expense ──")
        try:
            amount = float(input("Amount ($): "))
            category = input("Category (e.g. Food, Transport): ")
            description = input("Description: ")
            date_str = input("Date (YYYY-MM-DD) [leave blank for today]: ").strip()
            date = date_str if date_str else None
            expense = self.tracker.add_expense(amount, category, description, date)
            print(f"✅ Added: {expense}")
        except ValueError:
            print("❌ Invalid amount.")

    def _view_expenses(self):
        print("\n── View Expenses ──")
        cat = input("Filter by category (leave blank for all): ").strip()
        month = input("Filter by month YYYY-MM (leave blank for all): ").strip()
        print("Sort by: (1) Date  (2) Amount  [default: 1]")
        sort_choice = input("Sort option: ").strip()
        sort_by = "amount" if sort_choice == "2" else "date"
        print("Order: (1) Descending  (2) Ascending  [default: 1]")
        order_choice = input("Order option: ").strip()
        ascending = order_choice == "2"
        expenses = self.tracker.list_expenses(
            category=cat or None, month=month or None,
            sort_by=sort_by, ascending=ascending)
        if not expenses:
            print("No expenses found.")
            return
        print(f"\n{'ID':<5} {'Date':<12} {'Category':<15} {'Amount':>8}  Description")
        print("─" * 60)
        for e in expenses:
            print(f"{e.expense_id:<5} {e.date:<12} {e.category:<15} ${e.amount:>7.2f}  {e.description}")
        print(f"\nTotal: ${sum(e.amount for e in expenses):.2f}  ({len(expenses)} entries)")

    def _edit_expense(self):
        print("\n── Edit Expense ──")
        try:
            eid = int(input("Expense ID to edit: "))
            expense = self.tracker.get_expense(eid)
            if not expense:
                print("❌ Expense not found.")
                return
            print(f"Current: {expense}")
            updates = {}
            val = input(f"New amount [${expense.amount}]: ").strip()
            if val: updates["amount"] = float(val)
            val = input(f"New category [{expense.category}]: ").strip()
            if val: updates["category"] = val
            val = input(f"New description [{expense.description}]: ").strip()
            if val: updates["description"] = val
            val = input(f"New date [{expense.date}]: ").strip()
            if val: updates["date"] = val
            if updates:
                self.tracker.edit_expense(eid, **updates)
                print(f"✅ Updated: {self.tracker.get_expense(eid)}")
            else:
                print("No changes made.")
        except ValueError:
            print("❌ Invalid input.")

    def _delete_expense(self):
        print("\n── Delete Expense ──")
        try:
            eid = int(input("Expense ID to delete: "))
            expense = self.tracker.get_expense(eid)
            if not expense:
                print("❌ Expense not found.")
                return
            confirm = input(f"Delete '{expense}'? (y/n): ").strip().lower()
            if confirm == "y":
                self.tracker.delete_expense(eid)
                print("✅ Expense deleted.")
        except ValueError:
            print("❌ Invalid ID.")

    def _set_budget(self):
        print("\n── Set Budget ──")
        try:
            category = input("Category: ")
            limit = float(input("Monthly budget limit ($): "))
            budget = self.tracker.set_budget(category, limit)
            print(f"✅ Budget set: {budget}")
        except ValueError:
            print("❌ Invalid amount.")

    def _remove_budget(self):
        print("\n── Remove Budget ──")
        category = input("Category: ")
        if self.tracker.remove_budget(category):
            print(f"✅ Budget for '{category.title()}' removed.")
        else:
            print("❌ No budget found for that category.")

    def _view_summary(self):
        print("\n── Summary ──")
        month = input("Month YYYY-MM (leave blank for all time): ").strip()
        summary = self.tracker.summary(month=month or None)
        print(f"\n📊 Period: {summary['period']}")
        print(f"   Total Spent: ${summary['total']:.2f}  ({summary['count']} transactions)\n")
        if summary["by_category"]:
            print(f"{'Category':<18} {'Spent':>9} {'Budget':>9} {'Remaining':>10}  Status")
            print("─" * 60)
            for cat, data in sorted(summary["by_category"].items()):
                limit = f"${data['limit']:.2f}" if data["limit"] else "  —"
                remaining = f"${data['remaining']:.2f}" if data["remaining"] is not None else "  —"
                status = "🔴 OVER" if data["over_budget"] else ("🟢 OK" if data["limit"] else "")
                print(f"{cat:<18} ${data['spent']:>8.2f} {limit:>9} {remaining:>10}  {status}")
        else:
            print("No expenses recorded.")

    def _exit(self):
        print("\nGoodbye!")
        raise SystemExit  

if __name__ == "__main__":
    ExpenseTrackerInterface().run()

File: expense_tracker.c
/*
 * Expense Tracker System - C Implementation
 * Features: Add/edit/delete expenses, categories, budgets, summary
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ─────────────────────────────────────────────
   Constants
───────────────────────────────────────────── */

#define MAX_EXPENSES   100
#define MAX_BUDGETS    20
#define MAX_STR        100
#define MAX_DATE       11   /* YYYY-MM-DD + null terminator */

/* ─────────────────────────────────────────────
   Structs (equivalent to Python classes)
───────────────────────────────────────────── */

typedef struct {
    int    id;
    float  amount;
    char   category[MAX_STR];
    char   description[MAX_STR];
    char   date[MAX_DATE];
} Expense;

typedef struct {
    char  category[MAX_STR];
    float limit;
} Budget;

typedef struct {
    Expense expenses[MAX_EXPENSES];
    int     expense_count;
    int     next_id;
    Budget  budgets[MAX_BUDGETS];
    int     budget_count;
} ExpenseTracker;

/* ─────────────────────────────────────────────
   Utility Functions
───────────────────────────────────────────── */

/* Convert string to title case e.g. "food" -> "Food" */
void to_title_case(char *str) {
    if (str[0] >= 'a' && str[0] <= 'z')
        str[0] -= 32;
}

/* Remove trailing newline from input */
void strip_newline(char *str) {
    int len = strlen(str);
    if (len > 0 && str[len - 1] == '\n')
        str[len - 1] = '\0';
}

/* Get today's date as YYYY-MM-DD */
void get_today(char *buf) {
    strcpy(buf, "2026-03-01");
}

/* ─────────────────────────────────────────────
   Tracker Initialization
───────────────────────────────────────────── */

void tracker_init(ExpenseTracker *t) {
    t->expense_count = 0;
    t->next_id       = 1;
    t->budget_count  = 0;
}

/* ─────────────────────────────────────────────
   Expense Functions
───────────────────────────────────────────── */

/* Add a new expense */
int add_expense(ExpenseTracker *t, float amount, const char *category,
                const char *description, const char *date) {
    if (t->expense_count >= MAX_EXPENSES) {
        printf("Max expenses reached.\n");
        return -1;
    }

    Expense *e = &t->expenses[t->expense_count];
    e->id     = t->next_id++;
    e->amount = amount;
    strncpy(e->category,    category,    MAX_STR - 1);
    strncpy(e->description, description, MAX_STR - 1);
    strncpy(e->date,        date,        MAX_DATE - 1);
    to_title_case(e->category);

    t->expense_count++;
    return e->id;
}

/* Find expense index by ID, returns -1 if not found */
int find_expense(ExpenseTracker *t, int id) {
    for (int i = 0; i < t->expense_count; i++)
        if (t->expenses[i].id == id)
            return i;
    return -1;
}

/* Edit an existing expense */
int edit_expense(ExpenseTracker *t, int id, float amount, const char *category,
                 const char *description, const char *date) {
    int idx = find_expense(t, id);
    if (idx == -1) return 0;

    Expense *e = &t->expenses[idx];
    if (amount > 0)           e->amount = amount;
    if (strlen(category) > 0) {
        strncpy(e->category, category, MAX_STR - 1);
        to_title_case(e->category);
    }
    if (strlen(description) > 0)
        strncpy(e->description, description, MAX_STR - 1);
    if (strlen(date) > 0)
        strncpy(e->date, date, MAX_DATE - 1);

    return 1;
}

/* Delete an expense by ID */
int delete_expense(ExpenseTracker *t, int id) {
    int idx = find_expense(t, id);
    if (idx == -1) return 0;

    /* Shift all expenses after idx one position to the left */
    for (int i = idx; i < t->expense_count - 1; i++)
        t->expenses[i] = t->expenses[i + 1];

    t->expense_count--;
    return 1;
}

/* Print all expenses, optionally filtered by month */
void list_expenses(ExpenseTracker *t, const char *month, int sort_by_amount, int ascending) {
    Expense *matches[MAX_EXPENSES];
    int count = 0;

    for (int i = 0; i < t->expense_count; i++) {
        if (strlen(month) == 0 || strncmp(t->expenses[i].date, month, 7) == 0)
            matches[count++] = &t->expenses[i];
    }

    if (count == 0) {
        printf("No expenses found.\n");
        return;
    }

    /* Bubble sort */
    for (int i = 0; i < count - 1; i++) {
        for (int j = 0; j < count - i - 1; j++) {
            int swap = 0;
            if (sort_by_amount)
                swap = ascending
                    ? matches[j]->amount > matches[j+1]->amount
                    : matches[j]->amount < matches[j+1]->amount;
            else
                swap = ascending
                    ? strcmp(matches[j]->date, matches[j+1]->date) > 0
                    : strcmp(matches[j]->date, matches[j+1]->date) < 0;
            if (swap) {
                Expense *tmp  = matches[j];
                matches[j]    = matches[j+1];
                matches[j+1]  = tmp;
            }
        }
    }

    printf("\n%-5s %-12s %-15s %8s  %s\n", "ID", "Date", "Category", "Amount", "Description");
    printf("--------------------------------------------------------------\n");

    float total = 0;
    for (int i = 0; i < count; i++) {
        Expense *e = matches[i];
        printf("%-5d %-12s %-15s $%7.2f  %s\n",
               e->id, e->date, e->category, e->amount, e->description);
        total += e->amount;
    }
    printf("\nTotal: $%.2f  (%d entries)\n", total, count);
}

/* ─────────────────────────────────────────────
   Budget Functions
───────────────────────────────────────────── */

/* Find budget index by category, returns -1 if not found */
int find_budget(ExpenseTracker *t, const char *category) {
    for (int i = 0; i < t->budget_count; i++)
        if (strcmp(t->budgets[i].category, category) == 0)
            return i;
    return -1;
}

/* Set or update a budget */
void set_budget(ExpenseTracker *t, const char *category, float limit) {
    int idx = find_budget(t, category);
    if (idx != -1) {
        t->budgets[idx].limit = limit;
        return;
    }
    if (t->budget_count >= MAX_BUDGETS) {
        printf("Max budgets reached.\n");
        return;
    }
    Budget *b = &t->budgets[t->budget_count++];
    strncpy(b->category, category, MAX_STR - 1);
    to_title_case(b->category);
    b->limit = limit;
}

/* Remove a budget */
int remove_budget(ExpenseTracker *t, const char *category) {
    int idx = find_budget(t, category);
    if (idx == -1) return 0;
    for (int i = idx; i < t->budget_count - 1; i++)
        t->budgets[i] = t->budgets[i + 1];
    t->budget_count--;
    return 1;
}

/* ─────────────────────────────────────────────
   Summary Report
───────────────────────────────────────────── */

void summary(ExpenseTracker *t, const char *month) {
    printf("\nPeriod: %s\n", strlen(month) > 0 ? month : "All Time");

    float total = 0;
    int   count = 0;

    char  cats[MAX_BUDGETS][MAX_STR];
    float spent[MAX_BUDGETS];
    int   cat_count = 0;

    for (int i = 0; i < t->expense_count; i++) {
        Expense *e = &t->expenses[i];
        if (strlen(month) > 0 && strncmp(e->date, month, 7) != 0)
            continue;

        total += e->amount;
        count++;

        int found = -1;
        for (int j = 0; j < cat_count; j++)
            if (strcmp(cats[j], e->category) == 0) { found = j; break; }

        if (found == -1) {
            strncpy(cats[cat_count], e->category, MAX_STR - 1);
            spent[cat_count] = e->amount;
            cat_count++;
        } else {
            spent[found] += e->amount;
        }
    }

    printf("Total Spent: $%.2f  (%d transactions)\n\n", total, count);

    if (cat_count == 0) { printf("No expenses recorded.\n"); return; }

    printf("%-18s %9s %9s %10s  Status\n", "Category", "Spent", "Budget", "Remaining");
    printf("--------------------------------------------------------------\n");

    for (int i = 0; i < cat_count; i++) {
        int bidx = find_budget(t, cats[i]);
        if (bidx != -1) {
            float remaining = t->budgets[bidx].limit - spent[i];
            const char *status = spent[i] > t->budgets[bidx].limit ? "OVER" : "OK";
            printf("%-18s $%8.2f $%8.2f $%9.2f  %s\n",
                   cats[i], spent[i], t->budgets[bidx].limit, remaining, status);
        } else {
            printf("%-18s $%8.2f %9s %10s\n", cats[i], spent[i], "-", "-");
        }
    }
}

/* ─────────────────────────────────────────────
   CLI Menu Actions
───────────────────────────────────────────── */

void menu_add(ExpenseTracker *t) {
    float amount;
    char  category[MAX_STR], description[MAX_STR], date[MAX_DATE + 1];

    printf("\n-- Add Expense --\n");
    printf("Amount ($): ");
    scanf("%f", &amount);
    getchar();

    printf("Category: ");
    fgets(category, MAX_STR, stdin);
    strip_newline(category);

    printf("Description: ");
    fgets(description, MAX_STR, stdin);
    strip_newline(description);

    printf("Date (YYYY-MM-DD) [leave blank for today]: ");
    fgets(date, sizeof(date), stdin);
    strip_newline(date);
    if (strlen(date) == 0) get_today(date);

    int id = add_expense(t, amount, category, description, date);
    if (id != -1)
        printf("Added expense #%d: $%.2f | %s | %s\n", id, amount, category, description);
}

void menu_view(ExpenseTracker *t) {
    char month[8] = "";
    int  sort_by_amount = 0, ascending = 0;
    char buf[4];

    printf("\n-- View Expenses --\n");
    printf("Filter by month YYYY-MM (leave blank for all): ");
    fgets(month, sizeof(month), stdin);
    strip_newline(month);

    printf("Sort by: (1) Date  (2) Amount [default 1]: ");
    fgets(buf, sizeof(buf), stdin);
    if (buf[0] == '2') sort_by_amount = 1;

    printf("Order: (1) Descending  (2) Ascending [default 1]: ");
    fgets(buf, sizeof(buf), stdin);
    if (buf[0] == '2') ascending = 1;

    list_expenses(t, month, sort_by_amount, ascending);
}

void menu_edit(ExpenseTracker *t) {
    int   id;
    float amount = 0;
    char  category[MAX_STR] = "", description[MAX_STR] = "", date[MAX_DATE] = "";
    char  buf[MAX_STR];

    printf("\n-- Edit Expense --\n");
    printf("Expense ID to edit: ");
    scanf("%d", &id);
    getchar();

    int idx = find_expense(t, id);
    if (idx == -1) { printf("Expense not found.\n"); return; }

    Expense *e = &t->expenses[idx];
    printf("Current: #%d | $%.2f | %s | %s | %s\n",
           e->id, e->amount, e->category, e->description, e->date);

    printf("New amount [$%.2f] (0 to skip): ", e->amount);
    scanf("%f", &amount);
    getchar();

    printf("New category [%s] (blank to skip): ", e->category);
    fgets(buf, MAX_STR, stdin); strip_newline(buf);
    if (strlen(buf) > 0) strncpy(category, buf, MAX_STR - 1);

    printf("New description [%s] (blank to skip): ", e->description);
    fgets(buf, MAX_STR, stdin); strip_newline(buf);
    if (strlen(buf) > 0) strncpy(description, buf, MAX_STR - 1);

    printf("New date [%s] (blank to skip): ", e->date);
    fgets(buf, MAX_DATE, stdin); strip_newline(buf);
    if (strlen(buf) > 0) strncpy(date, buf, MAX_DATE - 1);

    edit_expense(t, id, amount, category, description, date);
    printf("Expense #%d updated.\n", id);
}

void menu_delete(ExpenseTracker *t) {
    int  id;
    char confirm[4];

    printf("\n-- Delete Expense --\n");
    printf("Expense ID to delete: ");
    scanf("%d", &id);
    getchar();

    int idx = find_expense(t, id);
    if (idx == -1) { printf("Expense not found.\n"); return; }

    Expense *e = &t->expenses[idx];
    printf("Delete #%d ($%.2f | %s)? (y/n): ", e->id, e->amount, e->description);
    fgets(confirm, sizeof(confirm), stdin);

    if (confirm[0] == 'y' || confirm[0] == 'Y') {
        delete_expense(t, id);
        printf("Expense deleted.\n");
    } else {
        printf("Cancelled.\n");
    }
}

void menu_set_budget(ExpenseTracker *t) {
    char  category[MAX_STR];
    float limit;

    printf("\n-- Set Budget --\n");
    printf("Category: ");
    fgets(category, MAX_STR, stdin);
    strip_newline(category);
    to_title_case(category);

    printf("Monthly budget limit ($): ");
    scanf("%f", &limit);
    getchar();

    set_budget(t, category, limit);
    printf("Budget set: %s -> $%.2f\n", category, limit);
}

void menu_remove_budget(ExpenseTracker *t) {
    char category[MAX_STR];

    printf("\n-- Remove Budget --\n");
    printf("Category: ");
    fgets(category, MAX_STR, stdin);
    strip_newline(category);
    to_title_case(category);

    if (remove_budget(t, category))
        printf("Budget for '%s' removed.\n", category);
    else
        printf("No budget found for '%s'.\n", category);
}

void menu_summary(ExpenseTracker *t) {
    char month[8] = "";
    printf("\n-- Summary --\n");
    printf("Month YYYY-MM (leave blank for all time): ");
    fgets(month, sizeof(month), stdin);
    strip_newline(month);
    summary(t, month);
}

/* ─────────────────────────────────────────────
   Main Menu Loop
───────────────────────────────────────────── */

void print_menu() {
    printf("\n");
    printf("╔══════════════════════════════════╗\n");
    printf("║       EXPENSE TRACKER            ║\n");
    printf("╠══════════════════════════════════╣\n");
    printf("║  1. Add Expense                  ║\n");
    printf("║  2. View Expenses                ║\n");
    printf("║  3. Edit Expense                 ║\n");
    printf("║  4. Delete Expense               ║\n");
    printf("║  5. Set Budget                   ║\n");
    printf("║  6. View Summary / Budget Status ║\n");
    printf("║  7. Remove Budget                ║\n");
    printf("║  0. Exit                         ║\n");
    printf("╚══════════════════════════════════╝\n");
}

int main() {
    ExpenseTracker tracker;
    tracker_init(&tracker);

    printf("\nWelcome to Expense Tracker!\n");

    char choice[4];
    while (1) {
        print_menu();
        printf("Choose an option: ");
        fgets(choice, sizeof(choice), stdin);
        strip_newline(choice);

        switch (choice[0]) {
            case '1': menu_add(&tracker);           break;
            case '2': menu_view(&tracker);          break;
            case '3': menu_edit(&tracker);          break;
            case '4': menu_delete(&tracker);        break;
            case '5': menu_set_budget(&tracker);    break;
            case '6': menu_summary(&tracker);       break;
            case '7': menu_remove_budget(&tracker); break;
            case '0':
                printf("\nGoodbye!\n");
                return 0;
            default:
                printf("Invalid option. Try again.\n");
        }
    }
}